### PR TITLE
Add HistoryODEIntegrator

### PR DESCRIPTION
**Pull request by devmotion**
This PR implements a reduced and cleaned version of https://github.com/JuliaDiffEq/DelayDiffEq.jl/pull/129 (some other features in that PR were implemented in the last days by other PRs).

The main idea is that by not using a full ODE integrator for the dummy integrator we can save a lot of allocations and type parameters. Initial tests show a huge decrease in test time and allocations compared with the master branch. However, I haven't tried to get rid of all type instabilities in `init`, so potentially further improvements are possible.

Some of the helper functions could/should be moved to OrdinaryDiffEq or DiffEqBase at some point, I guess.

### CODE REVIEW

**Comment by devmotion** [comment]
On Travis the total time it takes to run the tests decreased by more than 50%: [from around 44 minutes (Julia 1.0) and 33 minutes (Julia 1.3)](https://travis-ci.org/JuliaDiffEq/DelayDiffEq.jl/builds/564905204) to [around 22 minutes (Julia 1.0) and 13 minutes (Julia 1.3)](https://travis-ci.org/JuliaDiffEq/DelayDiffEq.jl/builds/565455026?utm_source=github_status&utm_medium=notification).

**Comment by devmotion** [comment]
The discrepancy between Julia 1.0 and Julia 1.3 can maybe explained by type instabilities.

Running
```julia
julia> @code_warntype init(DDEProblemLibrary.prob_dde_constant_1delay_ip, MethodOfSteps(Tsit5()));
```
shows that there are no type instabilities on Julia 1.3 whereas there are 6 type instabilities on Julia 1.1.1. I guess, some (maybe all?) of them can be fixed in similar ways as discussed in https://github.com/JuliaDiffEq/DelayDiffEq.jl/pull/129#issue-298605657. The problem of inferring the integrator options is gone, but the types of `p`, `resid`, `tdir`, `sol`, `f`, and `cache` cannot be inferred. By some tricks it was possible to infer at least the types of `p`, `resid`, and `tdir` in https://github.com/JuliaDiffEq/DelayDiffEq.jl/pull/129, but I haven't added these in this PR to keep it as clean as possible. I guess, those hacks could be added in a separate PR.

File: src/utils.jl [hunk 1]
```diff
@@ -190,110 +255,97 @@ function Base.sizehint!(sol::DESolution, alg, tspan, tstops, saveat;
...
+  # obtain cache (we alias uprev2 and uprev)
+  ode_cache = OrdinaryDiffEq.alg_cache(alg.alg, ode_u, rate_prototype, uEltypeNoUnits,
+                                       uBottomEltypeNoUnits, tTypeNoUnits, ode_uprev,
+                                       ode_uprev, ode_f, t, zero(dt), reltol, p, calck,
+                                       Val{isinplace(prob)})
+
+  # build dense interpolation of history
+  if iscomposite(alg)
+    ode_alg_choice = Int[]
+    ode_id = OrdinaryDiffEq.CompositeInterpolationData(ode_f, ode_timeseries, ode_ts, ode_ks,
```
Comment (ChrisRackauckas) [inline src/utils.jl]: we might only need these interpolations

**Comment by ChrisRackauckas** [comment]
having to do all of the options handling worries me, and I think we might only need the interpolations to do this correctly? The history function only needs to know how to get values, so if we do that and update it as necessary, it seems like we'd be good and we could drop the second integrator?

**Comment by devmotion** [comment]
> having to do all of the options handling worries me

Which options do you mean?

> The history function only needs to know how to get values, so if we do that and update it as necessary, it seems like we'd be good and we could drop the second integrator?

The whole purpose of the second integrator is to be able to extrapolate and to interpolate in the current integration interval. Currently the cache is also used for dispatching on which interpolation is actually used and to get the tableaus necessary for interpolation. So I don't see how we could get rid of this? Or is there a shortcut possible somewhere in OrdinaryDiffEq? But even in that case we would need the tableau?

**Comment by ChrisRackauckas** [comment]
>Which options do you mean?

All of the keyword arguments.

>The whole purpose of the second integrator is to be able to extrapolate and to interpolate in the current integration interval. Currently the cache is also used for dispatching on which interpolation is actually used and to get the tableaus necessary for interpolation. So I don't see how we could get rid of this? Or is there a shortcut possible somewhere in OrdinaryDiffEq? But even in that case we would need the tableau?

Currently the cache is used to dispatch because that's our dispatching system, but if we made the interpolants actually be different types then we could probably clean a bunch of stuff up, and as a side effect the dispatching for them would be independent of the integrator. Other than the `add_steps!` portions it would run completely automatically.

**Comment by devmotion** [comment]
> Currently the cache is used to dispatch because that's our dispatching system, but if we made the interpolants actually be different types then we could probably clean a bunch of stuff up, and as a side effect the dispatching for them would be independent of the integrator. Other than the add_steps! portions it would run completely automatically.

Yes, I think that could work.

> All of the keyword arguments.

My ideal would be if the DDEIntegrator is a struct that just wraps an ODEIntegrator, a history, and a fixed point solver. Then hopefully most arguments could just be forwarded.

**Pull request merged**